ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -169,32 +279,598 @@ module Json = struct
+            num - 1
+        | _ ->
+            num
+      in
+      match num with 0 -> method_name | _ -> method_name ^ string_of_int num
+    in
+    let params_in_msg msg =
+      if msg.msg_session then
+        session_id :: msg.msg_params
+      else
+        msg.msg_params
+    in
+    obj.messages
+    |> List.rev_map (fun msg ->
+           let of_message (num, params) =
+             let base_assoc_list =
+               [
+                 ("method_name", `String msg.msg_name)
+               ; ("class_name", `String obj.name)
+               ; ("class_name_exported", `String (snake_to_camel obj.name))
+               ; ( "method_name_exported"
+                 , `String (method_name_exported obj msg num)
+                 )
+               ; ("description", desc_of_msg msg ctor_fields)
+               ; ("result", of_result obj msg)
+               ; ("params", params)
+               ; ("errors", of_errors msg.msg_errors)
+               ; ("has_error", `Bool (msg.msg_errors <> []))
+               ; ("async", `Bool msg.msg_async)
+               ]
+             in
+             `O (add_session_info obj.name msg.msg_name @ base_assoc_list)
+           in
+           params_in_msg msg |> of_param_groups obj.name |> List.map of_message
+       )
+    |> List.concat
+
+  let of_option ty =
+    let name, _ = string_of_ty_with_enums ty () in
+    `O [("type", `String name); ("func_partial_type", `String (get_fp_type ty))]
+
+  let of_options types =
+    types
+    |> List.filter_map (fun ty ->
+           match ty with Option ty -> Some ty | _ -> None
+       )
+    |> List.map of_option
+
   let xenapi objs =
     List.map
       (fun obj ->
+        let obj_name = snake_to_camel obj.name in
+        let name_internal = String.uncapitalize_ascii obj_name in
         let fields = Datamodel_utils.fields_of_obj obj in
         let types = List.map (fun field -> field.ty) fields in
         let modules =
           match obj.messages with [] -> `Null | _ -> modules_of_types types
         in
         let base_assoc_list =
           [
-            ("name", `String (snake_to_camel obj.name))
+            ("name", `String obj_name)
+          ; ("name_internal", `String name_internal)
           ; ("description", `String (String.trim obj.description))
           ; ( "fields"
             , `A (get_event_snapshot obj.name @ List.map of_field fields)
             )
           ; ("modules", modules)
+          ; ("messages", `A (messages_of_obj obj))
+          ; ("option", `A (of_options types))
           ]
         in
         let assoc_list = base_assoc_list @ get_event_session_value obj.name in
-        (String.lowercase_ascii obj.name, `O assoc_list)
+        ( (String.lowercase_ascii obj.name, `O assoc_list)
+        , (obj_name ^ "Record", fields)
+        )
       )
       objs
+    |> List.split
+
+  let of_api_message_or_error info =
+    let snake_to_camel (s : string) : string =
```

Comment:
Are there multiple definitions of this function? I believe this should be a top-level function.

Reply:
No. Here it should be lowercased first, then capitalized here. I renamed it to `low_then_camel`.